File: 20-SEPT-19/Producer-Consumer.java
import java.util.concurrent.Semaphore;

class Q {
    
    int item = 0;

 
    static Semaphore semCon = new Semaphore(0);

    static Semaphore semProd = new Semaphore(1);

    
    void get() {
        try {
            semCon.acquire();
        } catch (InterruptedException e) {
            System.out.println("InterruptedException caught");
        }
            // this.item--;
        System.out.println("Consumer consumed item : " + item);
        semProd.release();
    }

    void put(int item) {
        try {
            semProd.acquire();
        } catch (InterruptedException e) {
            System.out.println("InterruptedException caught");
        }

        this.item = item; 

        System.out.println("Producer produced item : " + this.item);
        semCon.release();
    }
}

class Producer implements Runnable {
    Q q;

    Producer(Q q) {
        this.q = q;
        new Thread(this, "Producer").start();
    }

    public void run() {
        try {
            for (int i = 0; i < 5; i++)
            q.put(i);
        } catch ( Exception e)
        {
            
        }
        
    }
}

class Consumer implements Runnable {
    Q q;

    Consumer(Q q) {
        this.q = q;
        new Thread(this, "Consumer").start();
    }

    public void run() {
        try {
            for (int i = 0; i < 5; i++)
                q.get();
        } catch (Exception e) {

        }
    }
}

class PC {
    public static void main(String args[]) {
        Q q = new Q();

        new Consumer(q);
        new Producer(q);
    }
}

File: 20-SEPT-19/Producer Consumer.py
#!/usr/bin/env python
# coding: utf-8

# In[13]:


import random, time, threading 
from threading import BoundedSemaphore ,Thread


# In[14]:


max_items = 5


# In[15]:


container = BoundedSemaphore(max_items)


# In[16]:


def producer(nloops):
    for i in range(nloops):
        time.sleep(random.randrange(2, 5))
        print(time.ctime(), end=": ")
        try:
            container.release()
            print("Produced an item.")
        except ValueError:
            print("Full, skipping.")
            
def consumer(nloops):
    for i in range(nloops):
        time.sleep(random.randrange(2, 5))
        print(time.ctime(), end=": ")   
        if container.acquire(False):
            print("Consumed an item.")
        else:
            print("Empty, skipping.")
            


# In[17]:


threads = []


# In[18]:


nloops = random.randrange(3, 6)


# In[19]:


print("Starting with %s items." % max_items)


# In[20]:


threads.append(Thread(target=producer, args=(nloops,)))
threads.append(Thread(target=consumer, args=(random.randrange(nloops, nloops+max_items+2),)))


# In[21]:


for thread in threads: 
    thread.start()
for thread in threads:
    thread.join()
print("All done.")


# In[ ]:




